hibernate-core/src/main/java/org/hibernate/query/criteria/internal/QueryStructure.java

```diff
@@ -286,52 +287,17 @@ private FromImplementor locateImplicitSelection() {
 		return implicitSelection;
 	}
 
-	@SuppressWarnings({ "unchecked" })
 	private void renderFromClause(StringBuilder jpaqlQuery, RenderingContext renderingContext) {
 		renderingContext.getClauseStack().push( Clause.FROM );
 
 		try {
 			jpaqlQuery.append( " from " );
-			String sep = "";
-			for ( Root root : getRoots() ) {
-				( (FromImplementor) root ).prepareAlias( renderingContext );
-				jpaqlQuery.append( sep );
-				sep = ", ";
-				jpaqlQuery.append( ( (FromImplementor) root ).renderTableExpression( renderingContext ) );
-			}
 
-			for ( Root root : getRoots() ) {
-				renderJoins( jpaqlQuery, renderingContext, root.getJoins() );
-				if ( root instanceof RootImpl ) {
-					Set<TreatedRoot> treats = ( (RootImpl) root ).getTreats();
-					for ( TreatedRoot treat : treats ) {
-						renderJoins( jpaqlQuery, renderingContext, treat.getJoins() );
-						renderFetches( jpaqlQuery, renderingContext, treat.getFetches() );
-					}
-				}
-				renderFetches( jpaqlQuery, renderingContext, root.getFetches() );
-			}
+			doRenderFrom( getRoots(), jpaqlQuery, renderingContext );
 
 			if ( isSubQuery ) {
 				if ( correlationRoots != null ) {
-					for ( FromImplementor<?, ?> correlationRoot : correlationRoots ) {
```

Comment:
IMO this should be reverted and we should specially handle the left join case instead as this kind of querying might result in worse performance for some cases. A subquery like `select ... from alias.elementCollection` will only use the join table in the SQL. AFAIU, the changes you are doing here will always use the owner entity table in the from clause and then join the join table.

Reply:
yeah, that is the gist of the original implementation plan. I agree we should consider the LEFT JOIN correlated root as a special case to boost performance, though the downside is the code might be harder to read and maintain.

Will do as you suggested. Again, this PR might involve lots of back and forth.

Reply:
After investigation, I found HHH-14197 has different root cause so I exclude that ticket outside this PR and will create its own PR in near future.

Updated as your suggestion. Thanks.